src/masternodes/icxorder.cpp

```diff
@@ -0,0 +1,248 @@
+#include <masternodes/icxorder.h>
+
+/// @attention make sure that it does not overlap with other views !!!
+const unsigned char CICXOrderView::ICXOrderCreationTx           ::prefix = '1';
+const unsigned char CICXOrderView::ICXOrderKey                  ::prefix = '2';
+const unsigned char CICXOrderView::ICXOrderStatus               ::prefix = 'B';
+const unsigned char CICXOrderView::ICXMakeOfferCreationTx       ::prefix = '3';
+const unsigned char CICXOrderView::ICXMakeOfferKey              ::prefix = '4';
+const unsigned char CICXOrderView::ICXSubmitDFCHTLCCreationTx   ::prefix = '5';
+const unsigned char CICXOrderView::ICXSubmitDFCHTLCKey          ::prefix = '6';
+const unsigned char CICXOrderView::ICXSubmitDFCHTLCStatus       ::prefix = 'C';
+const unsigned char CICXOrderView::ICXSubmitEXTHTLCCreationTx   ::prefix = '7';
+const unsigned char CICXOrderView::ICXSubmitEXTHTLCKey          ::prefix = '8';
+const unsigned char CICXOrderView::ICXClaimDFCHTLCCreationTx    ::prefix = '9';
+const unsigned char CICXOrderView::ICXClaimDFCHTLCKey           ::prefix = '0';
+const unsigned char CICXOrderView::ICXCloseOrderCreationTx      ::prefix = 'A';
+
+const int CICXOrder::DEFAULT_EXPIRY = 2880;
+const int CICXOrder::TYPE_INTERNAL = 1;
+const int CICXOrder::TYPE_EXTERNAL = 2;
+const int CICXOrder::STATUS_OPEN = 0;
+const int CICXOrder::STATUS_CLOSED = 1;
+const int CICXOrder::STATUS_EXPIRED = 2;
+
+const int CICXMakeOffer::STATUS_OPEN = 0;
+
+const int CICXSubmitDFCHTLC::DEFAULT_TIMEOUT = 100;
+const int CICXSubmitDFCHTLC::STATUS_OPEN = 0;
+const int CICXSubmitDFCHTLC::STATUS_CLAIMED = 1;
+const int CICXSubmitDFCHTLC::STATUS_REFUNDED = 2;
+
+const int CICXSubmitEXTHTLC::STATUS_OPEN = 0;
+const int CICXSubmitEXTHTLC::STATUS_EXPIRED = 1;
+
+
+std::unique_ptr<CICXOrderView::CICXOrderImpl> CICXOrderView::GetICXOrderByCreationTx(const uint256 & txid) const
+{
+    auto order = ReadBy<ICXOrderCreationTx,CICXOrderImpl>(txid);
+    if (order)
+        return MakeUnique<CICXOrderImpl>(*order);
+    return (nullptr);
+}
+
+ResVal<uint256> CICXOrderView::ICXCreateOrder(const CICXOrderImpl& order)
+{
+    //this should not happen, but for sure
+    if (GetICXOrderByCreationTx(order.creationTx)) {
+        return Res::Err("order with creation tx %s already exists!", order.creationTx.GetHex());
+    }
+
+    OrderKey key({CICXOrder::STATUS_OPEN,{order.idToken,order.chain}}, order.creationTx);
+    WriteBy<ICXOrderCreationTx>(order.creationTx, order);
+    WriteBy<ICXOrderKey>(key, 1);
+    WriteBy<ICXOrderStatus>(StatusKey(order.creationHeight+order.expiry,order.creationTx),CICXOrder::STATUS_EXPIRED);
```

Comment:
Why we have status expired on creation?

Reply:
This key in db is basically on which block height the order expires. I could save `1` in value but I wanted to be semantically correct so I set expired status which will become on that height.